collector/spec.go

```diff
@@ -0,0 +1,28 @@
+package collector
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Interface defines how a collector implementation should look like.
+type Interface interface {
+	// Boot is used to initial and register the collector. Boot must be allowed to
+	// be called multiple times and thus must be idempotent. Usual implementations
+	// could make use of sync.Once. See also
+	// https://godoc.org/github.com/prometheus/client_golang/prometheus#Register.
+	Boot(ctx context.Context) error
+	// Collect should align with the monitoring system's implementation
+	// requirements. In this case Prometheus. See also
+	// https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector.
+	// The difference here is that this specific interface provides additional
+	// error handling capabilities.
+	CollectWithError(ch chan<- prometheus.Metric) error
```

Comment:
Should be `Collect` and not return error? It isn't aligned with godoc.

Reply:
The godoc is outdated. We talked with Joe and we wanted our collectors to be able to return errors so we have nicer implementations there. This here is right IMO. We cannot call it `Collect` WITH have the error because the Prometheus interface already defines `Collect` WITHOUT error.

Reply:
Why does it have to implement Prometheus interface?